File: output/rust/tests/37_context_reentrant.rs

// Test: Context Stack Reentrancy
// Validates that nested interface calls maintain separate contexts


use std::collections::HashMap;

#[derive(Clone, Debug)]
struct ContextReentrantTestFrameEvent {
    message: String,
    parameters: std::collections::HashMap<String, String>,
}

impl ContextReentrantTestFrameEvent {
    fn new(message: &str) -> Self {
        Self {
            message: message.to_string(),
            parameters: std::collections::HashMap::new(),
        }
    }
    fn with_parameters(message: &str, parameters: std::collections::HashMap<String, String>) -> Self {
        Self { message: message.to_string(), parameters }
    }
}

struct ContextReentrantTestFrameContext {
    event: ContextReentrantTestFrameEvent,
    _return: Option<Box<dyn std::any::Any>>,
    _data: std::collections::HashMap<String, Box<dyn std::any::Any>>,
}

impl ContextReentrantTestFrameContext {
    fn new(event: ContextReentrantTestFrameEvent, default_return: Option<Box<dyn std::any::Any>>) -> Self {
        Self {
            event,
            _return: default_return,
            _data: std::collections::HashMap::new(),
        }
    }
}

#[derive(Clone)]
struct ContextReentrantTestCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<ContextReentrantTestFrameEvent>,
}

impl ContextReentrantTestCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum ContextReentrantTestStateContext {
    Ready,
    Empty,
}

impl Default for ContextReentrantTestStateContext {
    fn default() -> Self {
        ContextReentrantTestStateContext::Ready
    }
}

pub struct ContextReentrantTest {
    _state_stack: Vec<(String, ContextReentrantTestStateContext)>,
    __compartment: ContextReentrantTestCompartment,
    __next_compartment: Option<ContextReentrantTestCompartment>,
    _context_stack: Vec<ContextReentrantTestFrameContext>,
}

impl ContextReentrantTest {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _context_stack: vec![],
            __compartment: ContextReentrantTestCompartment::new("Ready"),
            __next_compartment: None,
        };
let __frame_event = ContextReentrantTestFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: ContextReentrantTestFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = ContextReentrantTestFrameEvent::new("<$");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = ContextReentrantTestFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = ContextReentrantTestFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &ContextReentrantTestFrameEvent) {
match self.__compartment.state.as_str() {
    "Ready" => self._state_Ready(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: ContextReentrantTestCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Ready" => ContextReentrantTestStateContext::Ready,
    _ => ContextReentrantTestStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = ContextReentrantTestFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    ContextReentrantTestStateContext::Ready => {}
    ContextReentrantTestStateContext::Empty => {}
}
    }

    pub fn outer(&mut self, x: i32) -> String {
let __e = ContextReentrantTestFrameEvent::new("outer");
match self.__compartment.state.as_str() {
            "Ready" => self._s_Ready_outer(&__e, x),
            _ => Default::default(),
        }
    }

    pub fn inner(&mut self, y: i32) -> String {
let __e = ContextReentrantTestFrameEvent::new("inner");
match self.__compartment.state.as_str() {
            "Ready" => self._s_Ready_inner(&__e, y),
            _ => Default::default(),
        }
    }

    pub fn get_both(&mut self, a: i32, b: i32) -> String {
let __e = ContextReentrantTestFrameEvent::new("get_both");
match self.__compartment.state.as_str() {
            "Ready" => self._s_Ready_get_both(&__e, a, b),
            _ => Default::default(),
        }
    }

    fn _state_Ready(&mut self, __e: &ContextReentrantTestFrameEvent) {
match __e.message.as_str() {
    _ => {}
}
    }

    fn _s_Ready_inner(&mut self, __e: &ContextReentrantTestFrameEvent, y: i32) -> String {
// Inner has its own context
// @@.y should be inner's param, not outer's
return format!("{}", /* @@.y */);
    }

    fn _s_Ready_outer(&mut self, __e: &ContextReentrantTestFrameEvent, x: i32) -> String {
// Set our return before calling inner
return "outer_initial".to_string();;

// Call inner - should NOT clobber our return
let inner_result = self.inner(/* @@.x - context params not implemented for Rust */ * 10);

// Our return should still be accessible
// Update it with combined result
return format!("outer:{},inner:{}", /* @@.x */, inner_result);
    }

    fn _s_Ready_get_both(&mut self, __e: &ContextReentrantTestFrameEvent, a: i32, b: i32) -> String {
// Test that we can access multiple params
let result_a = self.inner(/* @@.a - context params not implemented for Rust */);
let result_b = self.inner(/* @@.b - context params not implemented for Rust */);
// After both inner calls, @@.a and @@.b should still be our params
return format!("a={},b={},results={}+{}", /* @@.a */, /* @@.b */, result_a, result_b);
    }
}


fn main() {
    println!("=== Test 37: Context Reentrant ===");
    let mut s = ContextReentrantTest::new();

    // Test 1: Simple nesting - outer calls inner
    let result = s.outer(5);
    let expected = "outer:5,inner:50";
    assert_eq!(result, expected, "Expected '{}', got '{}'", expected, result);
    println!("1. outer(5) = '{}'", result);

    // Test 2: Inner alone
    let result = s.inner(42);
    assert_eq!(result, "42", "Expected '42', got '{}'", result);
    println!("2. inner(42) = '{}'", result);

    // Test 3: Multiple inner calls, params preserved
    let result = s.get_both(10, 20);
    let expected = "a=10,b=20,results=10+20";
    assert_eq!(result, expected, "Expected '{}', got '{}'", expected, result);
    println!("3. get_both(10, 20) = '{}'", result);

    println!("PASS: Context reentrant works correctly");
}

File: output/go/tests/37_context_reentrant.go

package main

import (
	"fmt"
	"os"
	"strconv"
)

type ContextReentrantTestFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type ContextReentrantTestFrameContext struct {
    _event  ContextReentrantTestFrameEvent
    _return any
    _data   map[string]any
}

type ContextReentrantTestCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *ContextReentrantTestFrameEvent
    parentCompartment *ContextReentrantTestCompartment
}

func newContextReentrantTestCompartment(state string) *ContextReentrantTestCompartment {
    return &ContextReentrantTestCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *ContextReentrantTestCompartment) copy() *ContextReentrantTestCompartment {
    nc := &ContextReentrantTestCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type ContextReentrantTest struct {
    _state_stack []*ContextReentrantTestCompartment
    __compartment *ContextReentrantTestCompartment
    __next_compartment *ContextReentrantTestCompartment
    _context_stack []ContextReentrantTestFrameContext
}

func NewContextReentrantTest() *ContextReentrantTest {
    s := &ContextReentrantTest{}
    s._state_stack = make([]*ContextReentrantTestCompartment, 0)
    s._context_stack = make([]ContextReentrantTestFrameContext, 0)
    s.__compartment = newContextReentrantTestCompartment("Ready")
    s.__next_compartment = nil
    __frame_event := ContextReentrantTestFrameEvent{_message: "$>", _parameters: nil}
    __ctx := ContextReentrantTestFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *ContextReentrantTest) __kernel(__e *ContextReentrantTestFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &ContextReentrantTestFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &ContextReentrantTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &ContextReentrantTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *ContextReentrantTest) __router(__e *ContextReentrantTestFrameEvent) {
    switch s.__compartment.state {
    case "Ready":
        s._state_Ready(__e)
    }
}

func (s *ContextReentrantTest) __transition(next *ContextReentrantTestCompartment) {
    s.__next_compartment = next
}

func (s *ContextReentrantTest) Outer(x int) string {
    __params := map[string]any{
        "x": x,
    }
    __e := ContextReentrantTestFrameEvent{_message: "Outer", _parameters: __params}
    __ctx := ContextReentrantTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextReentrantTest) Inner(y int) string {
    __params := map[string]any{
        "y": y,
    }
    __e := ContextReentrantTestFrameEvent{_message: "Inner", _parameters: __params}
    __ctx := ContextReentrantTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextReentrantTest) DeeplyNested(z int) string {
    __params := map[string]any{
        "z": z,
    }
    __e := ContextReentrantTestFrameEvent{_message: "DeeplyNested", _parameters: __params}
    __ctx := ContextReentrantTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextReentrantTest) GetBoth(a int, b int) string {
    __params := map[string]any{
        "a": a,
        "b": b,
    }
    __e := ContextReentrantTestFrameEvent{_message: "GetBoth", _parameters: __params}
    __ctx := ContextReentrantTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextReentrantTest) _state_Ready(__e *ContextReentrantTestFrameEvent) {
    if __e._message == "DeeplyNested" {
        z := __e._parameters["z"].(int)
        _ = z
        // 3 levels deep
        outer_result := s.Outer(z)
        s._context_stack[len(s._context_stack)-1]._return = "deep:" + strconv.Itoa(z) + "," + outer_result
    } else if __e._message == "GetBoth" {
        a := __e._parameters["a"].(int)
        _ = a
        b := __e._parameters["b"].(int)
        _ = b
        // Test that we can access multiple params
        result_a := s.Inner(a)
        result_b := s.Inner(b)
        // After both inner calls, a and b should still be our params
        s._context_stack[len(s._context_stack)-1]._return = "a=" + strconv.Itoa(a) + ",b=" + strconv.Itoa(b) + ",results=" + result_a + "+" + result_b
    } else if __e._message == "Inner" {
        y := __e._parameters["y"].(int)
        _ = y
        // Inner has its own context
        s._context_stack[len(s._context_stack)-1]._return = strconv.Itoa(y)
    } else if __e._message == "Outer" {
        x := __e._parameters["x"].(int)
        _ = x
        // Set our return before calling inner
        s._context_stack[len(s._context_stack)-1]._return = "outer_initial"

        // Call inner - should NOT clobber our return
        inner_result := s.Inner(x * 10)

        // Our return should still be accessible
        // Update it with combined result
        s._context_stack[len(s._context_stack)-1]._return = "outer:" + strconv.Itoa(x) + ",inner:" + inner_result
    }
}

func main() {
	fmt.Println("=== Test 37: Context Reentrant ===")
	sm := NewContextReentrantTest()

	// Test 1: Simple nesting - outer calls inner
	result := sm.Outer(5)
	expected := "outer:5,inner:50"
	if result != expected {
		fmt.Printf("FAIL: Expected '%s', got '%s'\n", expected, result)
		os.Exit(1)
	}
	fmt.Printf("1. Outer(5) = '%s'\n", result)

	// Test 2: Inner alone
	result = sm.Inner(42)
	if result != "42" {
		fmt.Printf("FAIL: Expected '42', got '%s'\n", result)
		os.Exit(1)
	}
	fmt.Printf("2. Inner(42) = '%s'\n", result)

	// Test 3: Deep nesting (3 levels)
	result = sm.DeeplyNested(3)
	expected = "deep:3,outer:3,inner:30"
	if result != expected {
		fmt.Printf("FAIL: Expected '%s', got '%s'\n", expected, result)
		os.Exit(1)
	}
	fmt.Printf("3. DeeplyNested(3) = '%s'\n", result)

	// Test 4: Multiple inner calls, params preserved
	result = sm.GetBoth(10, 20)
	expected = "a=10,b=20,results=10+20"
	if result != expected {
		fmt.Printf("FAIL: Expected '%s', got '%s'\n", expected, result)
		os.Exit(1)
	}
	fmt.Printf("4. GetBoth(10, 20) = '%s'\n", result)

	fmt.Println("PASS: Context reentrant works correctly")
}
